src/ngx_http_lua_exten.c

```diff
@@ -0,0 +1,35 @@
+
+/*
+ * Copyright (C) Yichun Zhang (agentzh)
+ */
+
+
+#ifndef NGX_LUA_NO_FFI_API
+
+
+#ifndef DDEBUG
+#define DDEBUG 0
+#endif
+#include "ddebug.h"
+
+
+#include "ngx_http_lua_util.h"
+
+
+int
+ngx_http_lua_ffi_req_get_ext(ngx_http_request_t *r, u_char *out, size_t *len)
+{
+    if (r->connection->fd == (ngx_socket_t) -1) {
+        return NGX_HTTP_LUA_FFI_BAD_CONTEXT;
+    }
+
+    *len = ngx_min(r->exten.len, *len);
+    ngx_memcpy(out, r->exten.data, *len);
```

Comment:
It seems that we can simply return the `r->exten.data` buffer to the Lua land because the `ffi.string()` API function will performance data copying anyway. It's a waste of CPU time to do an extra data copying.